Remove "bower.json" and only create if necessary

Turbo87 opened the pull request:
"if necessary" being if a blueprint specifically wants to install Bower dependencies.

This PR removes the need for Ember CLI to lazily install Bower during `ember new`, since the `bower.json` file is now gone in the `app` and `addon` blueprints.

twokul commented on lib/tasks/bower-install.js:
```diff
@@ -1,32 +1,48 @@
   }
 
   installBower() {
+    logger.info('Installing "bower" via NPM into: %s', cliPath);
```
what's `cliPath` here? global or `node_modules/ember-cli`?

Turbo87 replied:
it depends ™

if you're in `ember new` it will be the global, if you're in `ember install` it will be the local

nathanhammond replied:
This somewhat falls in line with our goals of using global package managers. Note that we don't trigger `--save`. @Turbo87 I'd prefer to _only_ do `npm install -g bower`...

Turbo87 replied:
@nathanhammond we have no control over the installed bower version in that case

nathanhammond replied:
@Turbo87 We can

1. Check to see if there is a global, and if it is above our version. If so, we use it.
2. If no global, install one, return.
3. If global and under our version, install into the current `cliPath`.

Note: I'd prefer to always dump this into our top-level `ember-cli` instead of into the project's copy. You'll counter that version matters local to the version of CLI and I'd agree, especially for `bower` since we interact with it programmatically instead of shelling out, and we'll make no changes.

Turbo87 replied:
There are several issues with what you're proposing since we're not shelling out to bower but requiring it instead. Also using global bower seems unrelated to this PR and can be discussed separately.

nathanhammond replied:
This conversation went exactly as I expected, I'm comfortable with things as they stand. 😜 

> You'll counter that version matters local to the version of CLI and I'd agree, especially for bower since we interact with it programmatically instead of shelling out, and we'll make no changes.

twokul commented on lib/tasks/bower-install.js:
```diff
@@ -37,32 +53,61 @@ class BowerInstallTask extends Task {
+
+    // if we are running "bower install" from "ember init" and there is
+    // no "bower.json" we return early
+    if (!this.hasBowerJson() && !savePackages) {
```
is `!savePackages` necessary here?

Turbo87 replied:
yes, otherwise `addBowerPackagesToProject()` won't actually add anything to `bower.json` if it doesn't exist yet

nathanhammond replied:
Note: rather than omitting the file we can do `Object.keys().length` of `dependencies` and `devDependencies` to determine if we need to install it. It's also less likely to break `ember-try`.

Turbo87 replied:
we could, but then again, why have an essentially empty file in the project?

twokul commented on tests/unit/tasks/bower-install-test.js:
```diff
@@ -2,31 +2,60 @@
+      td.when(task.importBower('path/to/bower')).thenReturn('ok');
+
+      return task.ensureBower().then(ok => {
+        expect(ok).to.equal('ok');
```
👌 is👌

Turbo87 replied:
🤷‍♂️

twokul commented:
restarting Appveyor

Turbo87 commented:
I seriously have no idea anymore what's wrong with AppVeyor. The code itself seems to work fine, but CI is :trollface:ing us...

nathanhammond commented on lib/tasks/bower-install.js:
```diff
@@ -37,32 +53,61 @@ class BowerInstallTask extends Task {
+    // resolve "bower" and install via NPM if necessary
+    // -> then check if a "bower.json" file exists and create if necessary
+    // -> then run the actual "bower install"
+    return this.ensureBower().then(() => this.ensureBowerJson()).then(() => {
```
```javascript
RSVP.all([this.ensureBower(), this.ensureBowerJson()])
  .then(...)
```

Turbo87 replied:
in that case you might get a `bower.json` file without the dependency if the `npm install bower` fails

nathanhammond replied:
Don't care too strongly. Can you promote the `.then` chain to the next line for clarity?

nathanhammond commented on blueprints/app/files/bower.json:
```diff
@@ -1,5 +0,0 @@
-{
```
I'm tempted to move this into a different generator rather than just omitting it. `bower init` is _way_ overkill for something that will never be run as a `bower` package.

```
$ bower install --save 'jquery'
bower cached        https://github.com/jquery/jquery-dist.git#3.1.1
bower validate      3.1.1 against https://github.com/jquery/jquery-dist.git#*
bower no-json       No bower.json file to save to, use bower init to create one
```

(For people manually installing `bower` packages instead of transitively via `ember install`.)

nathanhammond replied:
(This comment only applies if we maintain the same approach of `hasBowerJson`.)

Turbo87 replied:
not sure how `bower init` is involved here

nathanhammond replied:
This is an ergonomics question, paired with https://github.com/ember-cli/ember-cli/pull/6698#discussion_r96703772 ... a blank `bower.json` file means that users can blissfully use `bower`. Otherwise the user is presented with the above message asking them to `bower init` which is _way_ more than they need.

Turbo87 replied:
The user will not be presented with any message!? We automatically create the file if we have to, though not through bower init.

stefanpenner replied:
let me know if i misunderstood this comment thread, but it sounds like:

* @nathanhammond is saying: using a warning a user would get when using `bower` cli directly themselves
*  @Turbo87 is saying: when ember cli invokes `bower`, it handles this case already.

If so, it seems reasonable to me that a user intending to once again use bower, is warned to first use `bower init`.

nathanhammond commented on lib/models/blueprint.js:
```diff
@@ -1017,7 +1017,7 @@ let Blueprint = CoreObject.extend({
       verbose: true,
       packages: packageNamesAndVersions,
-      installOptions,
+      installOptions: installOptions || { save: true },
```
This should be `merge`d. (I'm aware that the original didn't merge and that this is simply moved.)

Turbo87 replied:
that would technically be a breaking API change

nathanhammond replied:
I'd call it a bugfix. 😜 YOLO?

stefanpenner replied:
Sounds reasonable as a `[BUGFIX]` commit, independent of this PR.

nathanhammond commented on lib/tasks/bower-install.js:
```diff
@@ -37,32 +53,61 @@ class BowerInstallTask extends Task {
+      return Promise.resolve();
+    }
+
+    let projectName = this.project.name();
```
- ✅ `in-repo-addon`
- ✅ `app`
- ✅ `addon` (since it will be the dummy app)

Also, not actually used or relevant.

stefanpenner replied:
vanity nit-pick: if this was `let name = this.project.name()` the `writeJson` bellow could use concise object syntax `{ name }`

Turbo87 commented:
@stefanpenner @rwjblue friendly reminder 😉

stefanpenner commented on lib/tasks/bower-install.js:
```diff
@@ -1,32 +1,48 @@
 
-    return resolve('bower', { basedir: cliPath })
+    return this.resolveBower()
       .catch(() => this.installBower()
```
I would prefer if we only installed bower if a specific error occurs. For example, I would hate for some other issue (SyntaxError or something), from accidentally causing `bower` to be installed

stefanpenner commented:
with https://github.com/ember-cli/ember-cli/pull/6698#discussion_r97881345 fixed, you have my 👍  The other is merely a nit-pick.

Turbo87 commented:
- AppVeyor is broken as always
- @nathanhammond seems to agree with the PR now
- addressed @stefanpenner's comments

anyone please confirm and press the big (currently not so green) merge button 😉 

(do **not** use @homu to merge this, as I've changed the base branch for this PR)

rwjblue commented on lib/tasks/bower-install.js:
```diff
@@ -1,32 +1,53 @@
-      .then(bowerPath => require(bowerPath));
+    return this.resolveBower()
+      .catch(error => {
+        if (error.message.indexOf('Cannot find module \'bower\'') === -1) {
```
Is the error message platform dependent or guaranteed?

Turbo87 replied:
tl;dr this is the best we can do for now... (see https://github.com/substack/node-resolve/commit/c622aefeb286e479d536601e30bb828e69f86ec3#commitcomment-20625328)

rwjblue commented:
👏
